lib/Cron/Workers/StatsNotifications/Worker.php

```diff
@@ -0,0 +1,136 @@
+    $this->mailer->sender = $this->mailer->getSenderNameAndAddress($this->constructSenderEmail());
+    foreach(self::getDueTasks() as $task) {
+      try {
+        $this->mailer->send($this->constructNewsletter($task), $settings['address']);
+      } catch(\Exception $e) {
+        if(WP_DEBUG) {
+          throw $e;
+        }
+      } finally {
+        $this->markTaskAsFinished($task);
+      }
+      CronHelper::enforceExecutionLimit($this->timer);
+    }
+  }
+
+  private function constructSenderEmail() {
+    $url_parts = parse_url(home_url());
+    $site_name = strtolower($url_parts['host']);
+    if(strpos($site_name, 'www.') === 0) {
+      $site_name = substr($site_name, 4);
+    }
+    return [
+      'address' => self::SENDER_EMAIL_PREFIX . $site_name,
+      'name' => self::SENDER_EMAIL_PREFIX . $site_name,
+    ];
+  }
+
+  public static function getDueTasks() {
+    $date = new Carbon();
+    return ScheduledTask::orderByAsc('priority')
+      ->orderByAsc('updated_at')
+      ->whereNull('deleted_at')
+      ->where('status', ScheduledTask::STATUS_SCHEDULED)
+      ->whereLte('scheduled_at', $date)
+      ->where('type', self::TASK_TYPE)
+      ->limit(Sending::RESULT_BATCH_SIZE)
+      ->findMany();
+  }
+
+  private function constructNewsletter(ScheduledTask $task) {
+    $newsletter = $this->getNewsletter($task);
+    $link = NewsletterLink::findTopLinkForNewsletter($newsletter);
+    $context = $this->prepareContext($newsletter, $link);
+    return [
+      'subject' => sprintf(_x('Stats for email %s', 'title of an automatic email containing statistics (newsletter open rate, click rate, etc)', 'mailpoet'), $newsletter->subject),
+      'body' => [
+        'html' => $this->renderer->render('emails/statsNotification.html', $context),
+      ],
+    ];
+  }
+
+  private function getNewsletter(ScheduledTask $task) {
+    $statsNotificationModel = $task->statsNotification()->findOne();
+    $newsletter = $statsNotificationModel->newsletter()->findOne();
+    if(!$newsletter) {
+      throw new \Exception('Newsletter not found');
+    }
+    return $newsletter
+      ->withSendingQueue()
+      ->withTotalSent()
+      ->withStatistics();
+  }
+
+  private function prepareContext(Newsletter $newsletter, NewsletterLink $link = null) {
+    $clicked = ($newsletter->statistics['clicked'] * 100) / $newsletter->total_sent;
+    $opened = ($newsletter->statistics['opened'] * 100) / $newsletter->total_sent;
+    $unsubscribed = ($newsletter->statistics['unsubscribed'] * 100) / $newsletter->total_sent;
+    $context =  [
+      'subject' => $newsletter->subject,
+      'preheader' => sprintf(_x(
+        '%1$s%% opens, %2$s%% clicks, %3$s%% unsubscribes in a nutshell.', 'newsletter open rate, click rate and unsubscribe rate', 'mailpoet'),
+        number_format($clicked, 2),
+        number_format($opened, 2),
+        number_format($unsubscribed, 2)
+      ),
+      $context['topLinkClicks'] = 0,
```

Comment:
It's the first time I see such a recursive array definition, had to trace it in mind to check it's okay. How about making it more readable?

Reply:
Sorry, this is my typo, I didn't mean to write it this way. It is a copy & paste error.